### ISSUE TITLE
do_patch fails with patches that creates new files

### ISSUE THREAD

**Issue by lars18th** [jb-alvarado/media-autobuild_suite#1277]
Hi,

I'm using a custom `ffmpeg_extra.sh` with some custom patches. The `.diff` files are local, and all seems to work without troubles until I added a patch that adds a new file. The diff file includes:
```
new file mode 100644
index 0000000..3f1c7f6
--- /dev/null
+++ b/libavcodec/test_filter.c
```
And the suite FAILS to apply this patch. Testing the patch inside a Linux machine the patch can be applied, so I suspect the problem with the suite is when adding files.

Please, can you check this?
Thank you!

**Comment by RiCONfru** [comment]
It doesn't work with `patch` because the created files aren't cleaned up on vcs reset. You must use `git am` instead.

**Comment by wiiaboo** [comment]
It doesn't work with `patch` because the created files aren't cleaned up on vcs reset. You must use `git am` instead.

**Comment by lars18th** [comment]
> It doesn't work with `patch` because the created files aren't cleaned up on vcs reset. You must use `git am` instead.

Hi @wiiaboo ,

That doesn't work at all! The `ab-suite-pre_configure.log` indicates the same (only chaging with 'git am' instead of with 'patch' ):
```
CFLAGS: -mthreads -mtune=generic -O2 -pipe
LDFLAGS: -pipe -static-libgcc -static-libstdc++
_pre_configure 
[93test-filter.patch(B[m
	Patch couldn't be applied with 'git am'. Continuing without patching.
```

Any idea? Have you tried by yourself to try to apply a patch that adds a file?

Regards.

PD: Please, reopen this issue and don't close it before resolve it. :wink:

**Comment by wiiaboo** [comment]
You should `git clean -xfd` ffmpeg-git before trying to apply the patches.

**Comment by wiiaboo** [comment]
It's not a suite issue, it's an inherent issue with using `patch` diffs instead of `git am`. The new files are not deleted with a `git reset`.

**Comment by lars18th** [comment]
> It's not a suite issue, it's an inherent issue with using `patch` diffs instead of `git am`. The new files are not deleted with a `git reset`.

OK, ok. Now I understand the problem. Thank you for the tip!

> You should git clean -xfd ffmpeg-git before trying to apply the patches.

But this `ffmpeg_extra.sh` raises a syntax error!
```
#!/bin/bash

touch custom_updated

_pre_configure(){
    git clean -xfd
    do_patch "https://patchwork.ffmpeg.org/patch/11111/mbox/ test.patch" am
}
```

CONSOLE:
```
/build/fail.var: line 2: BASH_REMATCH: readonly variable
/build/fail.var: line 3534: syntax error near unexpected token `}'
/build/fail.var: line 3534: `}'
This is required for other packages, so this script will exit.
```

**Comment by wiiaboo** [comment]
No idea about that, if you did not change anything in the suite. @1480c1 ?

**Comment by lars18th** [comment]
> No idea about that, if you did not change anything in the suite. @1480c1 ?

Sure! Using the last version with a "git clone".
At time, my solution (to apply my custom patch) it's to add a command `rm -rf "$LOCALBUILDDIR/ffmpeg-git/myfile.c"` before the `do_patch` line.

In any case, please, update the documentation to describe:
1. The problem with the patches that adds files.
2. The requirement to use a filename when using patches from the patchwork.

Regards.

**Comment by lars18th** [comment]
And @wiiaboo , please _reopen_ the issue.
Thank you!

**Comment by wiiaboo** [comment]
If this is an issue from you misusing the _extra.sh functionality it's still not a suite issue.

**Comment by wiiaboo** [comment]
> * The problem with the patches that adds files.
> 
> * The requirement to use a filename when using patches from the patchwork.

Again, this is not a suite issue. I'm not going to document every single broken thing with using the suite as it's supposed to be used.

**Comment by wiiaboo** [comment]
The rule of thumb is, if it's broken whether it's through the suite, it's not a suite issue and therefore doesn't need to be documented in the suite. If you want to make a general FAQ of broken general things in the wiki, that's fine.

**Comment by lars18th** [comment]
> The rule of thumb is, if it's broken whether it's through the suite or directly in mintty, it's not a suite issue and therefore doesn't need to be documented in the suite. If you want to make a general FAQ of broken general things in the wiki, that's fine.

OK. I'll do for the "patches that adds new files".

But, this is a Copy&Paste for the **official** README file of the Suite:
https://github.com/jb-alvarado/media-autobuild_suite#custom-patches
```
Custom Patches
...
Example Script: /build/ffmpeg_extra.sh for ffmpeg-git
...
_pre_configure(){
    #
    # Apply a patch from ffmpeg's patchwork site.
    do_patch "https://patchwork.ffmpeg.org/patch/12563/mbox/" am
...
```

And as you know, and you commented here, the line `do_patch "https://patchwork.ffmpeg.org/patch/12563/mbox/" am` doesn't work.
With all respect, why you say "this is not a suite issue" ?

**Comment by lars18th** [comment]
> If you want to make a general FAQ of broken general things in the wiki, that's fine.

Done! https://github.com/jb-alvarado/media-autobuild_suite/wiki/Custom-Patches
For the "patches that adds new files". :wink:

**Comment by RiCONfru** [comment]
Because it needs specific hacks to support it, like I just did with c80f098c41752c1f480a1f013a3eaf8858d86f38.

**Comment by wiiaboo** [comment]
Because it needs specific hacks to support it, like I just did with c80f098c41752c1f480a1f013a3eaf8858d86f38.

**Comment by lars18th** [comment]
> Because it needs specific hacks to support it, like I just did with [c80f098](https://github.com/jb-alvarado/media-autobuild_suite/commit/c80f098c41752c1f480a1f013a3eaf8858d86f38).

OK. So just _modify_ the Homepage. The current official guide is outdated.

**Comment by 1480c1** [comment]
As far as I remember, only the aom one might actually work, the rest probably dont since I didn't really put much effort in confirming if they work

**Comment by lars18th** [comment]
> As far as I remember, only the aom one might actually work, the rest probably dont since I didn't really put much effort in confirming if they work

NO! I'm using it a lot for `ffmpeg_extra.sh`, and for this reason I report back this.
Please, commit the PR.

**Comment by 1480c1** [comment]
I mean for the examples on the readme

**Comment by lars18th** [comment]
> I mean for the examples on the readme

Ahh! Sorry!

Ok, I only suggest to update the example for the `ffmpeg_extra.sh'  with the change of the filename download.

Thank you for this amazing suite! :+1:

**Comment by wiiaboo** [comment]
There's no fucking need to update the example because it should be fixed by https://github.com/jb-alvarado/media-autobuild_suite/commit/c80f098c41752c1f480a1f013a3eaf8858d86f38 already (also fixed typo just now in a95df2e).

You got your hack, enough with the multi-thread discussion about a non-fucking issue, ffs.

**Comment by lars18th** [comment]
> There's no fucking need to update the example because it should be fixed by [c80f098](https://github.com/jb-alvarado/media-autobuild_suite/commit/c80f098c41752c1f480a1f013a3eaf8858d86f38) already (also fixed typo just now in [a95df2e](https://github.com/jb-alvarado/media-autobuild_suite/commit/a95df2ea0c89bf0efd220d572c04fe29d7f47b2d)).

Thank you for these two updates! :+1: 

> You got your hack, enough with the multi-thread discussion about a non-fucking issue, ffs.

I'm sorry, it's not my idea to offend you. I just want to contribute.

**Comment by wiiaboo** [comment]
I can't reproduce your syntax error issue and your patch doesn't apply on ffmpeg master either.

Avoid using `git clean` because it also removes suite files.
Use the new `safe_git_clean` I implemented in a0bea0780c56943a7c699fd5b5d77e4e0f5b1a55 instead, if you absolutely need to use it. You can test what would be deleted with `--dry-run` extra option.

> I'm sorry, it's not my idea to offend you. I just want to contribute.

I'm not offended. I just hate repeating myself.

**Comment by lars18th** [comment]
Hi @wiiaboo ,

> I can't reproduce your syntax error issue and your patch doesn't apply on ffmpeg master either.
> 
> Avoid using `git clean` because it also removes suite files.
> Use the new `safe_git_clean` I implemented in [a0bea07](https://github.com/jb-alvarado/media-autobuild_suite/commit/a0bea0780c56943a7c699fd5b5d77e4e0f5b1a55) instead, if you absolutely need to use it. You can test what would be deleted with `--dry-run` extra option.

I updated the suite, and then using my "stock" `ffmpeg_extra.sh` it seems to work. Thank you! :+1: 

> > I'm sorry, it's not my idea to offend you. I just want to contribute.
> 
> I'm not offended. I just hate repeating myself.

So, all issues are now fixed. :wink:
